base/abstractarray.jl

```diff
@@ -1945,111 +1945,34 @@ pushfirst!(A, a, b, c...) = pushfirst!(pushfirst!(A, c...), a, b)
-    h = hash(last(r), h)
-end
-
-function hash(a::AbstractArray{T}, h::UInt) where T
-    # O(1) hashing for types with regular step
-    if isa(a, AbstractRange) && isa(RangeStepStyle(a), RangeStepRegular)
-        return hash_range(a, h)
-    end
-
-    h += hashaa_seed
-    h += hash(size(a))
-
-    state = start(a)
-    done(a, state) && return h
-    x1, state = next(a, state)
-    done(a, state) && return hash(x1, h)
-    x2, state = next(a, state)
-    done(a, state) && return hash(x2, hash(x1, h))
-
-    # Check whether the array is equal to a range, and hash the elements
-    # at the beginning of the array as such as long as they match this assumption
-    # This needs to be done even for non-RangeStepRegular types since they may still be equal
-    # to RangeStepRegular values (e.g. 1.0:3.0 == 1:3)
-    if isa(a, AbstractVector) && applicable(-, x2, x1)
-        n = 1
-        local step, laststep, laststate
-        while true
-            # If overflow happens with entries of the same type, a cannot be equal
-            # to a range with more than two elements because more extreme values
-            # cannot be represented. We must still hash the two first values as a
-            # range since they can always be considered as such (in a wider type)
-            if isconcretetype(T)
-                try
-                    step = x2 - x1
-                catch err
-                    isa(err, OverflowError) || rethrow(err)
-                    break
-                end
-                # If true, wraparound overflow happened
-                sign(step) == cmp(x2, x1) || break
-            else
-                applicable(-, x2, x1) || break
-                # widen() is here to ensure no overflow can happen
-                step = widen(x2) - widen(x1)
-            end
-            n > 1 && !isequal(step, laststep) && break
-            n += 1
-            x1 = x2
-            laststep = step
-            laststate = state
-            done(a, state) && break
-            x2, state = next(a, state)
-        end
-
-        h = hash(first(a), h)
-        h += hashr_seed
-        # Always hash at least the two first elements as a range (even in case of overflow)
-        if n < 2
-            h = hash(2, h)
-            h = hash(x2, h)
-            done(a, state) && return h
-            x1 = x2
-            x2, state = next(a, state)
-        else
-            h = hash(n, h)
-            h = hash(x1, h)
-            done(a, laststate) && return h
+# A simple wrapper over checkbounds to support Reversed and Take iterator wrappers.
+# We assume that the original index iterators have optimized methods, so just ditch
+# the Reverse and Take wrappers and examine the original index iterator as a whole.
+_hashcheckbounds(A, I) = checkbounds(A, I)
+# But we cannot define those here, due to bootstrapping. They're defined later in iterators.jl.
+function hashndistinct(A, I, n, hx)
+    _hashcheckbounds(A, I)
+    seen = Set()
```

Comment:
Would probably be better to use a Vector for this since the number of items is small, and it will avoid hashing them multiple times.

Reply:
Yeah, I was thinking of using a specialized 3-vector to avoid allocations entirely, but figured we  could decide on the direction and exact behavior first before optimizing it further.